File: VirtualLab_Python/SynDiag_Python/diagnostics/Tokalab/Diag_FluxLoops.py
# -*- coding: utf-8 -*-
"""
Created on Tue Jun  3 17:38:38 2025

@author: ricca
"""

import numpy as np
import os
from scipy.interpolate import griddata
import matplotlib.pyplot as plt

class Diag_FluxLoops:
    def __init__(self):
        self.R = None  # Horizontal coordinate
        self.Z = None  # Vertical coordinate
        self.psi = None  # Measured magnetic flux
        self.sigma_psi = None # Uncertainty for Measured magnetic flux
        self.unit = None  # Unit of measurement
        self.config = {}  # Configuration parameters
        self.ideal = {}  # Ideal (noise-free) measurements

    def measure(self, equi):
        
        points = np.column_stack((equi.geo.grid.Rg.flatten(), equi.geo.grid.Zg.flatten()))

        # Interpolazione con griddata (equiv. interp2 Matlab)
        psi = griddata(points, equi.psi.flatten(), (self.R, self.Z), method='linear')

        # Save ideal (noise-free) signal
        self.ideal["psi"] = psi
        
        # Add noise
        noise_abs = np.random.normal(0, self.config.get("noise_random_absolute_intensity", 0), size=np.shape(psi.ravel()))
        noise_prop = np.random.normal(0, np.abs(psi) * self.config.get("noise_random_proportional_intensity", 0))

        self.psi = psi + noise_abs + noise_prop
        self.sigma_psi = np.sqrt(self.config.get("noise_random_absolute_intensity", 0)**2 +
                                    (np.abs(psi) * self.config.get("noise_random_proportional_intensity", 0))**2)
        self.unit = "Wb/rad"

    def upload(self, configuration=1):
        if configuration == 1:
            self.config["configuration"] = 1
            from scipy.io import loadmat
            module_path = os.path.abspath(__file__)
            module_path = os.path.dirname(module_path)
            module_path = os.path.join(module_path,"diagnostics_data","FluxLoopsData_config_1.mat")
            data = loadmat(module_path)
            
            self.R = np.squeeze(data["R"])
            self.Z = np.squeeze(data["Z"])

            self.config["noise_random_absolute_intensity"] = 0
            self.config["noise_random_proportional_intensity"] = 0
            
    #### Plotting function
    def plot_geo(self):
        # Scatter plot of geometry points (R, Z)
        plt.plot(self.R, self.Z, '.', markersize=16)
    
        # Show grid with both major and minor lines
        plt.grid(visible=True, which='both')
    
        # Axis labels
        plt.xlabel("R")
        plt.ylabel("Z")
    
        # Display the plot
        plt.show()
    
    def plot_meas(self):
        # Plot psi as a function of index
        plt.plot(self.psi, '.', markersize=16)
    
        # Show grid with both major and minor lines
        plt.grid(visible=True, which='both')
    
        # Axis labels
        plt.xlabel("#")
        plt.ylabel(r"$\psi$ [Wb/rad]")  # LaTeX formatting
    
        # Display the plot
        plt.show()
        
    def plot_StandAlone(self):
    
        # Plot ideal psi values as blue dots
        plt.plot(self.ideal["psi"].T, '.b', markersize=16)
    
        # Overlay actual psi values as red circles
        plt.plot(self.psi.T, 'or', linewidth=1.2)
    
        # Grid and formatting
        plt.grid(visible=True, which='both')
        plt.xlabel("#")
        plt.ylabel(r"$\psi$ [Wb/rad]")  # LaTeX for Greek letter psi
    
        # Display the plot
        plt.show()
    

File: VirtualLab_MatLab/SynDiag_MATLAB/diagnostics/Diag_FluxLoops.m
%% tokalab diagnostics

classdef Diag_FluxLoops

    properties

        R % Horixontal coordinate
        Z % Vertical coordinate

        psi % Measured Flux

        sigma_psi % Associated uncertainty for Measured Flux

        unit % Unit Measure

        config % contains the various information such as noise, etc.

        ideal % contains the measurements without the noise

    end

    methods

        function obj = measure(obj,equi)

            R_equi = equi.geo.grid.Rg;
            Z_equi = equi.geo.grid.Zg;

            psi_equi = equi.psi;

            psi = interp2(R_equi,Z_equi,psi_equi,obj.R,obj.Z);

            obj.ideal.psi = psi;

            % noise absolute
            noise_abs = normrnd(0,obj.config.noise_random_absolute_intensity,size(obj.ideal.psi));

            % noise proportional
            noise_prop = normrnd(0,abs(obj.ideal.psi).*obj.config.noise_random_proportional_intensity);

            % real measurement
            obj.psi = obj.ideal.psi + noise_abs + noise_prop;

            %associated uncertainty
            obj.sigma_psi = sqrt(obj.config.noise_random_absolute_intensity.^2 +...
                (abs(obj.ideal.psi).*obj.config.noise_random_proportional_intensity).^2);

            obj.unit = "Wb/rad";

        end

        function obj = Upload(obj,configuration,machine)

            % default configuration
            if nargin<2
                configuration = 1;
                machine = "TokaLab";
            elseif nargin < 3
                machine = "TokaLab";
            end

            %%%%%%%%%%%% TokaLab Configuration %%%%%%%%%%%%%%%%%%%%%%%%%%%%

            if machine == "TokaLab"

                if configuration == 1

                    obj.config.configuration = 1;

                    load("FluxLoopsData_TokaLab_config_1.mat")

                    obj.R = R;
                    obj.Z = Z;

                    obj.config.noise_random_absolute_intensity = 0;
                    obj.config.noise_random_proportional_intensity = 0;

                end
            end

            %%%%%%%%%%%% JET-like Configuration %%%%%%%%%%%%%%%%%%%%%%%%%%%%

            if machine == "JET-like"

                if configuration == 1

                    obj.config.configuration = 1;

                    load("FluxLoopsData_JETlike_config_1.mat")

                    obj.R = R;
                    obj.Z = Z;

                    obj.config.noise_random_absolute_intensity = 0;
                    obj.config.noise_random_proportional_intensity = 0;

                end
            end

            %%%%%%%%%%%% DTT-like Configuration %%%%%%%%%%%%%%%%%%%%%%%%%%%%

            if machine == "DTT-like"

                if configuration == 1

                    obj.config.configuration = 1;

                    load("FluxLoopsData_DTTlike_config_1.mat")

                    obj.R = R;
                    obj.Z = Z;

                    obj.config.noise_random_absolute_intensity = 0;
                    obj.config.noise_random_proportional_intensity = 0;

                end
            end
        end


            %% Plotting Functions

            function plot_geo(obj)

                plot(obj.R,obj.Z,'.','MarkerSize',16)
                grid on
                grid minor
                xlabel("R")
                ylabel("Z")

            end

            function plot_meas(obj)

                plot(obj.psi,'.','MarkerSize',16)
                grid on
                grid minor
                xlabel("#")
                ylabel("\psi [Wb/rad]")

            end

            function plot_StandAlone(obj)

                hold off
                plot(obj.ideal.psi,'.b','MarkerSize',16)
                hold on
                plot(obj.psi,'or','LineWidth',1.2)
                grid on
                grid minor
                xlabel("#")
                ylabel("\psi [Wb/rad]")

            end
    end
end



